pennylane/ops/arithmetic/adjoint.py

```diff
@@ -0,0 +1,435 @@
+
+    @property
+    def grad_recipe(self):
+        r"""tuple(Union(list[list[float]], None)) or None: Gradient recipe for the
+        parameter-shift method.
+
+        This is a tuple with one nested list per operation parameter. For
+        parameter :math:`\phi_k`, the nested list contains elements of the form
+        :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+        term, resulting in a gradient recipe of
+
+        .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+        If ``None``, the default gradient recipe containing the two terms
+        :math:`[c_0, a_0, s_0]=[1/2, 1, \pi/2]` and :math:`[c_1, a_1,
+        s_1]=[-1/2, 1, -\pi/2]` is assumed for every parameter.
+        """
+        return self.base.grad_recipe
+
+    def get_parameter_shift(self, idx):
+        r"""Multiplier and shift for the given parameter, based on its gradient recipe.
+
+        Args:
+            idx (int): parameter index within the operation
+
+        Returns:
+            list[[float, float, float]]: list of multiplier, coefficient, shift for each term in the gradient recipe
+
+        Note that the default value for ``shift`` is None, which is replaced by the
+        default shift :math:`\pi/2`.
+        """
+        return self.base.get_parameter_shift(idx)
+
+    @property
+    def parameter_frequencies(self):
+        r"""Returns the frequencies for each operator parameter with respect
+        to an expectation value of the form
+        :math:`\langle \psi | U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p})|\psi\rangle`.
+
+        These frequencies encode the behaviour of the operator :math:`U(\mathbf{p})`
+        on the value of the expectation value as the parameters are modified.
+        For more details, please see the :mod:`.pennylane.fourier` module.
+
+        Returns:
+            list[tuple[int or float]]: Tuple of frequencies for each parameter.
+            Note that only non-negative frequency values are returned.
+
+        **Example**
+
+        >>> op = qml.CRot(0.4, 0.1, 0.3, wires=[0, 1])
+        >>> op.parameter_frequencies
+        [(0.5, 1), (0.5, 1), (0.5, 1)]
+
+        For operators that define a generator, the parameter frequencies are directly
+        related to the eigenvalues of the generator:
+
+        >>> op = qml.ControlledPhaseShift(0.1, wires=[0, 1])
+        >>> op.parameter_frequencies
+        [(1,)]
+        >>> gen = qml.generator(op, format="observable")
+        >>> gen_eigvals = qml.eigvals(gen)
+        >>> qml.gradients.eigvals_to_frequencies(tuple(gen_eigvals))
+        (1.0,)
+
+        For more details on this relationship, see :func:`.eigvals_to_frequencies`.
+        """
+        return self.base.parameter_frequencies
+
+
+def adjoint(fn, lazy=True):
+    """Create a function that applies the adjoint of the provided operation or template.
+
+    This transform can be used to apply the adjoint of an arbitrary sequence of operations.
```

Comment:
Actually not.  It only wraps things that were processed into `tape.operations`. So that might be a problem. Wrapping everything in `tape.queue` will open up a bunch of complications.